### DISCUSSION TITLE
Automatic migration is not available

### DISCUSSION THREAD

**Discussion by damonearp** [apache/airflow#18912] [Q&A] [answered]
After trying to upgrade to 2.2.0 DB Migration fails. Logs report several task_instances and a task_fail without corresponding dag_run entries.

I logged into the postgres instance and ran
```
postgres=# select task_id from task_instance where dag_id not in (select dag_id from dag_run);
 task_id 
---------
(0 rows)
```

That's curious. Maybe it's via the job table?
```
postgres=# select task_id from task_instance where job_id not in (select job_id from job);
 task_id 
---------
(0 rows)

postgres=# select task_id from task_instance where job_id in (select job_id from job where dag_id not in (select dag_id from dag_run));
 task_id 
---------
(0 rows)
```
Any help resolving this issue would be nice. Thanks!

--- 

Upgrade from 2.1.4 to 2.2.0
Image: apache/airflow:2.2.0-python3.9
Entrypoint: <unchanged>
CMD: [ db, upgrade]

Logs
```
/home/airflow/.local/lib/python3.9/site-packages/airflow/configuration.py:361 DeprecationWarning: The dag_concurrency option in [core] has been renamed to max_active_tasks_per_dag - the old setting has been used, but please update your config.
/home/airflow/.local/lib/python3.9/site-packages/airflow/configuration.py:361 DeprecationWarning: The processor_poll_interval option in [scheduler] has been renamed to scheduler_idle_sleep_time - the old setting has been used, but please update your config.
/home/airflow/.local/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/base.py:3571 SAWarning: Predicate of partial index idx_dag_run_queued_dags ignored during reflection
/home/airflow/.local/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/base.py:3571 SAWarning: Predicate of partial index idx_dag_run_running_dags ignored during reflection
DB: postgresql+psycopg2:///?service=airflow
[2021-10-12 16:16:37,889] {db.py:815} ERROR - Automatic migration is not available
[2021-10-12 16:16:37,889] {db.py:817} ERROR - The task_instance table has 13 rows without a corresponding dag_run row. You must manually correct this problem (possibly by deleting the problem rows).
[2021-10-12 16:16:37,892] {db.py:817} ERROR - The task_fail table has 1 row without a corresponding dag_run row. You must manually correct this problem (possibly by deleting the problem rows).
```

# Update 2021-11-02
Just tried 2.1.4 to 2.2.1 and the migration code moved the bad rows to another table. Inspecting the rows I cannot tell what's wrong with those entries. 

After the upgrade I saw a banner on the WebUI
> Airflow found incompatible data in the task_instance table in the metadatabase, and has moved them to _airflow_moved__2_2__task_instance during the database migration to upgrade. Please inspect the moved data to decide whether you need to keep them, and manually drop the _airflow_moved__2_2__task_instance table to dismiss this warning.

Only thing I can think of is this table lacks a `run_id` column, was that the missing link?

**Comment by nicozingarelli** [comment]
Anyone have any insights on this topic? I am getting the same behavior. I've tried running python 3.8 and 3.9 locally to see if the issue was more package related but it does not seem to make a difference

**Comment by uranusjr** [comment, marked as answer]
The `run_id` column was added in 2.2.0 as an explicit link between `dag_run` and `task_instance`. The rows are moved because the upgrade script cannot find a matching DAG run for them (those are also called "dangling"). This is usually caused by you deleting a DAG run from the web UI but Airflow failed to clean up those task instances, in which case those can be deleted safely. If you really want to keep those entries, you'll need to manually create DAG runs for those task instances to link to, and insert those rows back into `task_instance` with an appropriate `run_id` you manually created.

**Reply by damonearp** [reply]
Ah so my mistake in the initial queries was matching the `task_instance` to a `dag` not to a `dag_run` instance. Thank you for the insight.
